src/main.rs

```diff
@@ -1,35 +1,690 @@
+        empty_box.open();
+    }
+
+    #[test]
+    fn filling_box() {
+        let filled_box = GiftBox::fill(Some(["Toys", "Candy", "Money"]));
+        assert_eq!(filled_box, GiftBox::Gifts(["Toys", "Candy", "Money"]));
+    }
+
+    #[test]
+    fn open_filled_box() {
+        let filled_box = GiftBox::fill(Some(vec![1, 3, 5, 7]));
+        assert_eq!(filled_box.open(), vec![1, 3, 5, 7]);
+    }
+
+    #[test]
+    fn filling_with_some() {
+        let some_filled_box = GiftBox::fill_keeping_some(Some("Chocolate"));
+        assert_eq!(some_filled_box.open(), Some("Chocolate"));
+    }
+
+    #[test]
+    fn filling_with_none() {
+        let empty_box: GiftBox<()> = GiftBox::fill(None);
+        assert_eq!(empty_box, GiftBox::Empty);
+    }
+
+    #[test]
+    fn wrap_gift_box_with_tag() {
+        let filled_box = GiftBox::fill(Some(["Toys", "Candy", "Money"]));
+        let tag = GiftTag::write(
+            "Bob".to_string(),
+            "Sally".to_string(),
+            "Happy Cake Day!".to_string()
+        );
+        let wrapped_box = filled_box.wrap(
+            Patterns::Polkadots,
+            true,
+            Some(tag)
+        );
+        assert_eq!(
+            wrapped_box,
+            {
+                GiftWrap {
+                    contents: {
+                        GiftBox::Gifts(["Toys", "Candy", "Money"])
+                    },
+                    pattern: Patterns::Polkadots,
+                    has_bow: true,
+                    tag: Some(
+                        GiftTag {
+                            recipient: "Bob".to_string(),
+                            sender: "Sally".to_string(),
+                            message: "Happy Cake Day!".to_string()
+                        }
+                    )
+                }
+            }
+        )
+    }
+
+    #[test]
+    fn wrap_and_unwrap_gift_box() {
+        let filled_box = GiftBox::fill(Some(["Toys", "Candy", "Money"]));
+        let tag = GiftTag::write(
+            "Bob".to_string(),
+            "Sally".to_string(),
+            "Happy Cake Day!".to_string()
+        );
+        let wrapped_box = filled_box.wrap(
+            Patterns::Polkadots,
+            true,
+            Some(tag)
+        );
+        let unwrapped_box = wrapped_box.unwrap();
+        assert_eq!(unwrapped_box, filled_box);
+    }
+
+    #[test]
+    fn wrap_and_open_gift_box_with_tag() {
+        let filled_box = GiftBox::fill(Some(["Toys", "Candy", "Money"]));
+        let tag = GiftTag::write(
+            "Bob".to_string(),
+            "Sally".to_string(),
+            "Happy Cake Day!".to_string()
+        );
+        let wrapped_box = filled_box.wrap(
+            Patterns::Polkadots,
+            true,
+            Some(tag)
+        );
+        assert_eq!(wrapped_box.unwrap().open(), ["Toys", "Candy", "Money"]);
+    }
+
+    #[test]
+    fn write_and_read_tag() {
+        let filled_box = GiftBox::fill(Some(["Toys", "Candy", "Money"]));
```

Comment:
this test should be able to isolate just the GiftTag::read function alone (or if you adjust it as commented above, then then you would test the Display trait by using `format!("{}", test_tag)`)

Reply:
I've redone this test and moved it to the gifttag tests in commit https://github.com/hypnoseal/giftbox/commit/92b666d99c2fe2af201453ca0f0097d7ea79bdbc